// checkly/dist/constructs/alert-channel-subscription.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AlertChannelSubscription = void 0;
const construct_1 = require("./construct");
const project_1 = require("./project");
/**
 * Creates an Alert Channel Subscription
 *
 * @remarks
 *
 * This class make use of the Alert Channel Subscriptions endpoints.
 */
class AlertChannelSubscription extends construct_1.Construct {
    /**
     * Constructs the Alert Channel Syscription instance
     *
     * @param logicalId unique project-scoped resource name identification
     * @param props alert channel subscription configuration properties
     */
    constructor(logicalId, props) {
        super(AlertChannelSubscription.__checklyType, logicalId);
        this.alertChannelId = props.alertChannelId;
        this.checkId = props.checkId;
        this.groupId = props.groupId;
        this.activated = props.activated;
        project_1.Session.registerConstruct(this);
    }
    synthesize() {
        return {
            alertChannelId: this.alertChannelId,
            checkId: this.checkId,
            groupId: this.groupId,
            activated: this.activated,
        };
    }
}
exports.AlertChannelSubscription = AlertChannelSubscription;
AlertChannelSubscription.__checklyType = 'alert-channel-subscription';
