// checkly/dist/constructs/maintenance-window.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MaintenanceWindow = void 0;
const construct_1 = require("./construct");
const project_1 = require("./project");
/**
 * Creates a Maintenance Window
 *
 * @remarks
 *
 * This class make use of the Maintenance Window endpoints.
 */
class MaintenanceWindow extends construct_1.Construct {
    /**
     * Constructs the Maintenance Window instance
     *
     * @param logicalId unique project-scoped resource name identification
     * @param props maintenance window configuration properties
     *
     * {@link https://checklyhq.com/docs/cli/constructs/#maintenancewindow Read more in the docs}
     */
    constructor(logicalId, props) {
        super(MaintenanceWindow.__checklyType, logicalId);
        this.name = props.name;
        this.tags = props.tags;
        this.startsAt = props.startsAt;
        this.endsAt = props.endsAt;
        this.repeatInterval = props.repeatInterval;
        this.repeatUnit = props.repeatUnit;
        this.repeatEndsAt = props.repeatEndsAt;
        project_1.Session.registerConstruct(this);
    }
    synthesize() {
        return {
            name: this.name,
            tags: this.tags,
            startsAt: this.startsAt,
            endsAt: this.endsAt,
            repeatInterval: this.repeatInterval,
            repeatUnit: this.repeatUnit,
            repeatEndsAt: this.repeatEndsAt,
        };
    }
}
exports.MaintenanceWindow = MaintenanceWindow;
MaintenanceWindow.__checklyType = 'maintenance-window';
