// checkly/dist/constructs/pagerduty-alert-channel.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PagerdutyAlertChannel = void 0;
const alert_channel_1 = require("./alert-channel");
const project_1 = require("./project");
/**
 * Creates an Pagerduty Alert Channel
 *
 * @remarks
 *
 * This class make use of the Alert Channel endpoints.
 */
class PagerdutyAlertChannel extends alert_channel_1.AlertChannel {
    /**
     * Constructs the Pagerduty Alert Channel instance
     *
     * @param logicalId unique project-scoped resource name identification
     * @param props Pagerduty alert channel configuration properties
     *
     * {@link https://checklyhq.com/docs/cli/constructs/#pagerdutyalertchannel Read more in the docs}
     */
    constructor(logicalId, props) {
        super(logicalId, props);
        this.account = props.account;
        this.serviceName = props.serviceName;
        this.serviceKey = props.serviceKey;
        project_1.Session.registerConstruct(this);
    }
    synthesize() {
        return {
            ...super.synthesize(),
            type: 'PAGERDUTY',
            config: {
                account: this.account,
                serviceName: this.serviceName,
                serviceKey: this.serviceKey,
            },
        };
    }
}
exports.PagerdutyAlertChannel = PagerdutyAlertChannel;
