// checkly/dist/constructs/phone-call-alert-channel.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PhoneCallAlertChannel = void 0;
const alert_channel_1 = require("./alert-channel");
const project_1 = require("./project");
/**
 * Creates a Phone Call Alert Channel
 *
 * @remarks
 *
 * This class make use of the Alert Channel endpoints.
 */
class PhoneCallAlertChannel extends alert_channel_1.AlertChannel {
    /**
     * Constructs the Phone Call Alert Channel instance
     *
     * @param logicalId unique project-scoped resource name identification
     * @param props Phone Call alert channel configuration properties
     *
     * {@link https://checklyhq.com/docs/cli/constructs-reference/#phonecallalertchannel Read more in the docs}
     * {@link https://checklyhq.com/docs/alerting/phone-calls/#supported-countries-and-regions
     *  List of supported countries}
     */
    constructor(logicalId, props) {
        super(logicalId, props);
        this.phoneNumber = props.phoneNumber;
        this.name = props.name;
        project_1.Session.registerConstruct(this);
    }
    synthesize() {
        return {
            ...super.synthesize(),
            type: 'CALL',
            config: {
                number: this.phoneNumber,
                name: this.name,
            },
        };
    }
}
exports.PhoneCallAlertChannel = PhoneCallAlertChannel;
