// checkly/dist/constructs/slack-alert-channel.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SlackAlertChannel = void 0;
const alert_channel_1 = require("./alert-channel");
const project_1 = require("./project");
/**
 * Creates an Slack Alert Channel
 *
 * @remarks
 *
 * This class make use of the Alert Channel endpoints.
 */
class SlackAlertChannel extends alert_channel_1.AlertChannel {
    /**
     * Constructs the Slack Alert Channel instance
     *
     * @param logicalId unique project-scoped resource name identification
     * @param props Slack alert channel configuration properties
     *
     * {@link https://checklyhq.com/docs/cli/constructs-reference/#smsalertchannel Read more in the docs}
     */
    constructor(logicalId, props) {
        super(logicalId, props);
        this.url = props.url;
        this.channel = props.channel;
        project_1.Session.registerConstruct(this);
    }
    synthesize() {
        return {
            ...super.synthesize(),
            type: 'SLACK',
            config: {
                url: this.url,
                channel: this.channel,
            },
        };
    }
}
exports.SlackAlertChannel = SlackAlertChannel;
