// checkly/dist/constructs/webhook-alert-channel.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.WebhookAlertChannel = void 0;
const alert_channel_1 = require("./alert-channel");
const project_1 = require("./project");
/**
 * Creates an Webhook Alert Channel
 *
 * @remarks
 *
 * This class make use of the Alert Channel endpoints.
 */
class WebhookAlertChannel extends alert_channel_1.AlertChannel {
    /**
     * Constructs the Webhook Alert Channel instance
     *
     * @param logicalId unique project-scoped resource name identification
     * @param props Webhook alert channel configuration properties
     *
     * {@link https://checklyhq.com/docs/cli/constructs/#webhookalertchannel Read more in the docs}
     */
    constructor(logicalId, props) {
        super(logicalId, props);
        this.name = props.name;
        this.webhookType = props.webhookType;
        this.url = props.url;
        this.template = props.template;
        this.method = props.method;
        this.headers = props.headers;
        this.queryParameters = props.queryParameters;
        project_1.Session.registerConstruct(this);
    }
    synthesize() {
        return {
            ...super.synthesize(),
            type: 'WEBHOOK',
            config: {
                name: this.name,
                webhookType: this.webhookType,
                url: this.url,
                template: this.template,
                method: this.method,
                headers: this.headers,
                queryParameters: this.queryParameters,
            },
        };
    }
}
exports.WebhookAlertChannel = WebhookAlertChannel;
