// checkly/dist/commands/login.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f) throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver)) throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
var _Login_instances, _Login_promptForLoginOrSignUp;
Object.defineProperty(exports, "__esModule", { value: true });
exports.selectAccount = void 0;
const open_1 = __importDefault(require("open"));
const chalk_1 = __importDefault(require("chalk"));
const baseCommand_1 = require("./baseCommand");
const prompts_1 = __importDefault(require("prompts"));
const config_1 = __importDefault(require("../services/config"));
const api = __importStar(require("../rest/api"));
const auth_1 = require("../auth");
const selectAccount = async (accounts, { onCancel }) => {
    if (accounts.length === 1) {
        return accounts[0];
    }
    const { selectedAccount } = await (0, prompts_1.default)({
        name: 'selectedAccount',
        type: 'select',
        choices: accounts.map(account => ({ title: account.name, value: account })),
        message: 'Which account do you want to use?',
    }, { onCancel });
    return selectedAccount;
};
exports.selectAccount = selectAccount;
class Login extends baseCommand_1.BaseCommand {
    constructor() {
        super(...arguments);
        _Login_instances.add(this);
        this._checkExistingCredentials = async () => {
            if (config_1.default.hasEnvVarsConfigured()) {
                this.warn('`CHECKLY_API_KEY` ' +
                    'or `CHECKLY_ACCOUNT_ID` environment variables are configured. You must delete them to use `npx checkly login`.');
                this.exit(0);
            }
            const hasValidCredentials = config_1.default.hasValidCredentials();
            if (hasValidCredentials) {
                const { setNewkey } = await (0, prompts_1.default)({
                    name: 'setNewkey',
                    type: 'confirm',
                    message: `You are currently logged in to "${config_1.default.data.get('accountName')}". Do you want to log out and log in to a different account?`,
                });
                !setNewkey && this.exit(0);
            }
        };
        this._isLoginSuccess = async () => {
            await api.validateAuthentication();
            this.log('Welcome to the Checkly CLI');
        };
    }
    async run() {
        await this._checkExistingCredentials();
        const mode = await __classPrivateFieldGet(this, _Login_instances, "m", _Login_promptForLoginOrSignUp).call(this);
        const onCancel = () => {
            this.error('Command cancelled.\n');
        };
        const authContext = new auth_1.AuthContext(mode);
        const { openUrl } = await (0, prompts_1.default)({
            name: 'openUrl',
            type: 'confirm',
            message: `Do you want to open a browser window to continue with ${mode === 'login' ? 'login' : 'sign up'}?`,
            initial: true,
        });
        if (!openUrl) {
            this.log(`Please open the following URL in your browser: \n\n${chalk_1.default.cyan(authContext.authenticationUrl)}`);
        }
        else {
            await (0, open_1.default)(authContext.authenticationUrl);
        }
        const { key, name } = await authContext.getAuth0Credentials();
        config_1.default.auth.set('apiKey', key);
        const { data: accounts } = await api.accounts.getAll();
        const selectedAccount = await (0, exports.selectAccount)(accounts, { onCancel });
        if (!selectedAccount) {
            this.warn('You must select a valid Checkly account name.');
            this.exit(1);
            return;
        }
        config_1.default.data.set('accountId', selectedAccount.id);
        config_1.default.data.set('accountName', selectedAccount.name);
        this.log(`Successfully logged in as ${chalk_1.default.cyan.bold(name)}`);
        await this._isLoginSuccess();
        this.exit(0);
    }
}
exports.default = Login;
_Login_instances = new WeakSet(), _Login_promptForLoginOrSignUp = async function _Login_promptForLoginOrSignUp() {
    const { mode } = await (0, prompts_1.default)({
        name: 'mode',
        type: 'select',
        message: 'Do you want to log in or sign up to Checkly?',
        choices: [{
                title: 'I want to log in with an existing Checkly account',
                value: 'login',
            }, {
                title: 'I want to sign up for a new Checkly account',
                value: 'signup',
            }],
    });
    return mode;
};
Login.hidden = false;
Login.description = 'Login to your Checkly account or create a new one.';
