// checkly/dist/commands/runtimes.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
const api = __importStar(require("../rest/api"));
const baseCommand_1 = require("./baseCommand");
class Runtimes extends baseCommand_1.BaseCommand {
    async run() {
        const { data: runtimes } = await api.runtimes.getAll();
        const output = runtimes
            .sort((a, b) => b.name.localeCompare(a.name))
            .map(r => {
            const { node, ...dependencies } = r.dependencies;
            return {
                name: r.name,
                description: r.description,
                stage: r.stage,
                nodeVersion: node,
                dependencies,
            };
        });
        this.log(JSON.stringify(output, null, 2));
    }
}
exports.default = Runtimes;
Runtimes.hidden = false;
Runtimes.description = 'List all supported runtimes and dependencies.';
