// checkly/dist/services/socket-client.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SocketClient = void 0;
const mqtt = __importStar(require("async-mqtt"));
const config_1 = __importDefault(require("../services/config"));
// @ts-ignore
const proxy_from_env_1 = require("proxy-from-env");
const tunnel_1 = require("tunnel");
const isHttps = (protocol) => protocol.startsWith('https');
class SocketClient {
    static connect() {
        const url = config_1.default.getMqttUrl();
        const accountId = config_1.default.getAccountId();
        const apiKey = config_1.default.getApiKey();
        const options = {
            reconnectPeriod: 0,
            username: accountId,
            password: apiKey,
        };
        // Replace wss with https so the get proxy url thing the env path
        const proxyUrlEnv = (0, proxy_from_env_1.getProxyForUrl)(url.replace('wss', 'https'));
        if (proxyUrlEnv) {
            const parsedProxyUrl = new URL(proxyUrlEnv);
            const isProxyHttps = isHttps(parsedProxyUrl.protocol);
            const proxy = {
                host: parsedProxyUrl.hostname,
                port: parsedProxyUrl.port,
                protocol: parsedProxyUrl.protocol,
            };
            if (parsedProxyUrl.username && parsedProxyUrl.password) {
                proxy.proxyAuth = `${proxyUrlEnv.username}:${proxyUrlEnv.password}`;
            }
            if (isProxyHttps) {
                options.wsOptions = {
                    agent: (0, tunnel_1.httpsOverHttps)({ proxy }),
                };
            }
            else {
                options.wsOptions = {
                    agent: (0, tunnel_1.httpsOverHttp)({ proxy }),
                };
            }
        }
        return mqtt.connectAsync(`${url}?authenticationScheme=userApiKey`, options, false);
    }
}
exports.SocketClient = SocketClient;
