// checkly/dist/rest/assets.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
// eslint-disable-next-line no-restricted-syntax
var AssetType;
(function (AssetType) {
    AssetType["LOG"] = "log";
    AssetType["CHECK_RUN_DATA"] = "check-run-data";
    AssetType["SCREENSHOT"] = "screenshot";
})(AssetType || (AssetType = {}));
class Assets {
    constructor(api) {
        this.api = api;
    }
    getLogs(region, key) {
        return this.getAssets(AssetType.LOG, region, key);
    }
    getCheckRunData(region, key) {
        return this.getAssets(AssetType.CHECK_RUN_DATA, region, key);
    }
    async getAssets(assetType, region, key) {
        const response = await this.api.get(`/next/assets/${assetType}/${region}/${encodeURIComponent(key)}`);
        return response.data;
    }
    async getAssetsLink(region, key) {
        const response = await this.api.get(`/next/assets/${AssetType.CHECK_RUN_DATA}/${region}/${encodeURIComponent(key)}/redirect?link=true`);
        return response.data;
    }
}
exports.default = Assets;
